design/external-authorization-http-support-design.md

```diff
@@ -0,0 +1,165 @@
+
+A new type, `GlobalHTTPExtAuth`, will be defined as part of the `ContourConfiguration`.
+A `GlobalHTTPExtAuth` config would define a global external authorization configuration for all the HTTP routes. 
+
+## Detailed Design
+
+### Contour Configuration changes
+By using global external authorization, an external authorization service can be configured in the Contour config file.
+This External authorization configuration will be used for all HTTP routes. 
+
+```yaml
+globalHTTPExtAuth:
+  # extensionService is the <namespace>/<name>
+  # of the ExtensionService we created in the
+  # previous step.
+  extensionService: projectcontour-auth/htpasswd
+  # failOpen is whether to allow requests through
+  # if there's an error.
+  failOpen: false
+  # Context is a set of key/value pairs that are sent to the authentication server in the check request.
+  context:
+    header1: value1
+    header2: value2
+    # ResponseTimeout configures maximum time to wait for a check response from the authorization server
+  responseTimeout: 1s
+
+```
+
+### HTTPS Override
+If an upstream has HTTPS enabled, the HTTPS configuration will take precedence. 
+i.e., if global HTTP external authorization is enabled and an upstream is configured with HTTPS without external authorization. Then that particular upstream would not use envoy's external authorization filter. 
+
+## Alternatives Considered
+
+### Envoy Filter Matching API 
+Having a separate external authorization configuration for each upstream is more desirable. We  can achieve this by leveraging Envoy's [Matching API]()
+Note: This feature is an Alpha release, and there are no guarantees that future changes will maintain backward compatibility. 
+
+Below is an example of how this would work. 
+
+Consider 2 HTTPProxy definitions defined below, both HTTP with external authorization filter enabled. 
+##### HTTP Proxy 1 
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: local.projectcontour.io
+spec:
+  virtualhost:
+    fqdn: local.projectcontour.io
+    authorization:
+      extensionRef:
+        name: htpasswd
+        namespace: projectcontour-auth
+      responseTimeout: 0.5s
+      authPolicy:
+        context:
+          header1: value1
+  routes:
+  - services:
+    - name: local.projectcontour.io
+      port: 80
+```
+
+##### HTTP Proxy 2
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: internal.projectcontour.io
+spec:
+  virtualhost:
+    fqdn: internal.projectcontour.io
+    tls:
+      secretName: ingress-conformance-echo
+    authorization:
+      extensionRef:
+        name: htpasswd
+        namespace: projectcontour-auth
+      responseTimeout: 1s
+      authPolicy:
+        context:
+          header2: value2
+  routes:
+  - services:
+    - name: internal.projectcontour.io
+      port: 80
+```
+
+With this proposal, contour will generate the envoy configuration snippet below. 
```

Comment:
I'm not entirely sure this example will work, I think the [ServerNameInput](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/matching/common_inputs/network/v3/network_inputs.proto#extensions-matching-common-inputs-network-v3-servernameinput) is for matching SNI. You would have to match on the `:authority` header

Haven't played around with the `Composite` and `ExecuteFilterAction` filters/actions so curious to see how that works

not a big deal to generate a working example here so we can omit it if needed, but if we do want to provide an example we should validate it works

Reply:
Also I think the TLS-vhost auth filter configuration would still go on the HTTPS listener and specific filter chain, rather than be combined with the HTTP one

Reply:
Fair, I should've validated it. 
I made some changes; the new config works with the `host` header and the `HttpRequestHeaderMatchInput` matching API. 

Also, agree with the TLS-vhost filter, which would be on the HTTPS listener. I added a note to make that clear.